pkg/maps/ctmap/ctmap.go

```diff
@@ -181,6 +183,106 @@ type GCFilter struct {
 	MatchIPs map[string]struct{}
 }
 
+func CreateCtKey(remoteAddr, localAddr string, proto u8proto.U8proto, ingress bool) (CtKey, error) {
+	ip, port, err := net.SplitHostPort(remoteAddr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid remote address '%s': %s", remoteAddr, err)
+	}
+
+	sIP := net.ParseIP(ip)
+	if sIP == nil {
+		return nil, fmt.Errorf("unable to parse IP %s", ip)
+	}
+
+	sport, err := strconv.ParseUint(port, 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse port string: %s", err)
+	}
+
+	localIp, localPort, err := net.SplitHostPort(localAddr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid local address '%s': %s", localAddr, err)
+	}
+
+	dIP := net.ParseIP(localIp)
+	if dIP == nil {
+		return nil, fmt.Errorf("unable to parse IP %s", localIp)
+	}
+
+	dport, err := strconv.ParseUint(localPort, 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse port string: %s", err)
+	}
+
+	if sIP.To4() != nil {
+		key := &CtKey4{
+			SourcePort: uint16(sport),
+			DestPort:   uint16(dport),
+			NextHeader: proto,
+			Flags:      TUPLE_F_OUT,
+		}
+		// CTmap has the addresses in the reverse order w.r.t. the original direction
+		copy(key.SourceAddr[:], dIP.To4())
+		copy(key.DestAddr[:], sIP.To4())
+		if ingress {
+			key.Flags = TUPLE_F_IN
+		}
+		return key.ToNetwork(), nil
+	}
+
+	key := &CtKey6{
+		SourcePort: uint16(sport),
+		DestPort:   uint16(dport),
+		NextHeader: proto,
+		Flags:      TUPLE_F_OUT,
+	}
+	// CTmap has the addresses in the reverse order w.r.t. the original direction
+	copy(key.SourceAddr[:], dIP.To16())
+	copy(key.DestAddr[:], sIP.To16())
+	if ingress {
+		key.Flags = TUPLE_F_IN
+	}
+	return key.ToNetwork(), nil
+}
+
+func Lookup(mapname string, remoteAddr, localAddr string, proto u8proto.U8proto, ingress bool) (*CtEntry, error) {
+	key, err := CreateCtKey(remoteAddr, localAddr, proto, ingress)
+	if err != nil {
+		return nil, err
+	}
+	if mapname == "" {
+		mapname = "global"
+	}
+	if _, ok := key.(CtKey); ok {
+		if proto == u8proto.TCP {
+			mapname = MapNameTCP4 + mapname
+		} else {
+			mapname = MapNameAny4 + mapname
+		}
+	} else {
+		if proto == u8proto.TCP {
+			mapname = MapNameTCP6 + mapname
+		} else {
+			mapname = MapNameAny6 + mapname
+		}
+	}
+
+	m := bpf.GetMap(mapname)
+	if m == nil {
+		// TODO: Figure out if the global maps ever need to be closed/reopened.
```

Comment:
My intent was to use the existing map registry to keep the ctmap open, so that each lookup would not need to bear the cost of opening and closing it, but this really works for the global map only, as this leaks fds for removed endpoints with local ct maps.

I've now changed this to:
- Add this new lookup functionality to `pkg/maps/ctmap/lookup.go`
- Add new `CloseLocalMaps()` that is called from `endpoint.LeaveLocked()` to close any opened local conntrack maps.

I don't readily see how to use the existing `maps()` function et al to get the protocol specific map. I'd like to leave tidying this up further for a follow-on PR.

Reply:
Ah, so:
* At least one lookup is guaranteed to happen,
* Furthermore the first lookup is guaranteed to open the map and leave it open,
* Then Close is guaranteed to remove a reference to the map.

And for global maps this will hopefully just open exactly one reference and for the local maps they won't be shared so each endpoint manages the reference to the fd in that case.

Seems fine at a glance, but perhaps a touch brittle. I'd welcome a follow-on PR to tidy this up a bit and make stronger guarantees here. Either way, I'm sure we'll find out sooner or later if the fd logic is wrong through fd open stats in testing clusters. SGTM for now.